File: Blink Comentado/PYTHON/blink.py
import gpiod  # Importa la biblioteca para interactuar con los pines GPIO a nivel de sistema
import time  # Importa la biblioteca para manejar los tiempos de espera y retardo

# Definición de la propiedad que representa el pin donde está conectado el LED
LED_PIN = 18  # El pin GPIO 18 controlará el encendido y apagado del LED

# Crea un objeto `chip` que representa el chip de control GPIO
# En términos de POO, este objeto es la instancia de la clase `Chip`, que interactúa con el hardware
chip = gpiod.Chip('gpiochip4')  # Inicializa el chip GPIO correspondiente (gpiochip4)

# Crea un objeto `led_line` que representa la línea GPIO correspondiente al pin LED_PIN
# Este objeto se comporta como una abstracción de la línea GPIO
led_line = chip.get_line(LED_PIN)  # Obtiene la línea GPIO del pin 18

# Solicita acceso a la línea GPIO configurándola como salida
# Aquí, `request()` es un método que configura la dirección del pin y establece quién lo está usando
led_line.request(consumer='LED', type=gpiod.LINE_REQ_DIR_OUT)  # Configura el pin 18 como salida

# Bucle que alterna el estado del LED (encendido/apagado) continuamente
try:
    while True:
        # Establece el valor de la línea en 1, lo que enciende el LED
        # Este método `set_value()` cambia el estado del objeto LED, similar a un método de clase
        led_line.set_value(1)
        time.sleep(1)  # Pausa por 1 segundo (equivalente a un método de retardo)

        # Establece el valor de la línea en 0, lo que apaga el LED
        led_line.set_value(0)
        time.sleep(1)  # Pausa por 1 segundo
except KeyboardInterrupt:
    # Manejo de la interrupción del teclado (CTRL + C)
    # Cuando se interrumpe, se libera la línea GPIO y se cierra el chip, similar a la destrucción del objeto
    led_line.release()  # Libera la línea GPIO (comportamiento similar a un método de limpieza/destructor)
    chip.close()  # Cierra el chip GPIO, liberando recursos

File: Blink Comentado/JS/blink.js
"use strict"; // Activa el modo estricto en JavaScript para una mejor gestión de errores y seguridad en la sintaxis

// Importa la biblioteca gpiox para controlar los pines GPIO como si fuera una clase externa que proporciona métodos para interactuar con hardware
const gpiox = require("@iiot2k/gpiox");

// Define el objeto `PIN` que actuará como un atributo de la clase, representando el pin GPIO a usar
const PIN = 18; // Pin GPIO donde se conectará el LED

// Define el objeto `INTERVAL` que actuará como una propiedad de la clase, representando el intervalo de parpadeo
const INTERVAL = 500; // Intervalo en milisegundos para alternar el estado del LED

// Inicializa el objeto GPIO como una salida, con un estado inicial de apagado (0)
// Este método se comporta como un constructor que configura las propiedades iniciales del objeto GPIO
gpiox.init_gpio(PIN, gpiox.GPIO_MODE_OUTPUT, 0);

// Define una propiedad de estado `ledState`, que controla si el LED está encendido o apagado
let ledState = 0;

// Método `setInterval` que funciona como un ciclo continuo dentro de la clase, alternando el estado del LED cada cierto intervalo
const intervalId = setInterval(() => {
    // Alterna el estado del LED usando una expresión ternaria; simula un método de "cambio de estado"
    ledState = (ledState === 0) ? 1 : 0; // Cambia el estado del LED (ON/OFF)
    
    // Llama al método que cambia el estado del pin GPIO, simulando el comportamiento de una clase que controla el hardware
    gpiox.set_gpio(PIN, ledState);

    // Verifica si el tiempo transcurrido ha superado los 10 segundos, detiene el parpadeo en ese caso
    // Esto actúa como un método de control del ciclo de vida del parpadeo
    if (Date.now() - startTime > 10000) { // Comprueba si han pasado 10 segundos (10000 ms)
        clearInterval(intervalId); // Detiene el ciclo de parpadeo al eliminar el intervalo
        gpiox.deinit_gpio(PIN); // Desinicializa el GPIO, liberando los recursos
        console.log("Parpadeo detenido y GPIO desinicializado."); // Imprime un mensaje indicando que se ha detenido
    }
}, INTERVAL);

// Guarda el tiempo de inicio en el objeto `startTime`, que será usado para controlar cuándo detener el parpadeo
const startTime = Date.now();
